configfile: "config.yaml"
include: "functions.smk" #Python functions

rule fastqc:
    input:
        get_bwa_map_input_fastqs
    output:
        "../results/fastqc/{sample}_fastqc.html",
        "../results/fastqc/{sample}_fastqc.zip"
    threads: 4
    conda:
        "envs/fastqc.yaml"
    log:
        "../results/logs/fastqc/{sample}.log"
    shell:
        "(fastqc -t {threads} -o ../results/fastqc/ {input}) > {log} 2>&1"

rule multiqc:
    input:
        '../results/fastqc/' #Gathers all FastQC results
    output:
        "../results/fastqc/multiqc_report.html"
    conda:
        "envs/multiqc.yaml"
    log:
        "../results/logs/fastqc/multiqc.log"
    shell:
        "(multiqc {input} -o ../results/fastqc/) > {log}"

rule bwa_map:
    input:
        ref="../data/references/hg38.fa",
        fastq = get_bwa_map_input_fastqs
    output:
        "../results/mapped_reads/{sample}.bam"
    params:
        rg=r"@RG\tID:{sample}\tSM:{sample}"   
    threads: 8
    conda:
        "envs/bwa.yaml"
    log:
        "../results/logs/bwa_mem/{sample}.log"
    shell:
        "(bwa mem -R '{params.rg}' -t {threads} {input.ref} {input.fastq} | samtools view -Sb - > {output}) > {log} 2>&1"

#BAM sorting with samtools
rule samtools_sort:
    input:
        "../results/mapped_reads/{sample}.bam"
    output:
        "../results/sorted_reads/{sample}.bam"
    conda:
        "envs/bwa.yaml"    
    shell:
        "samtools sort -T ../results/sorted_reads/{wildcards.sample} -O bam {input} > {output}"

#BAM indexing with samtools
rule samtools_index:
    input:
        "../results/sorted_reads/{sample}.bam"
    output:
        "../results/sorted_reads/{sample}.bam.bai"
    conda:
        "envs/bwa.yaml"    
    shell:
        "samtools index {input}"

#Basic vcf call with bcftools -- Only for WES/WGS data
'''
rule bcftools_call:
    input:
        ref="../data/references/hg38.fa",
        bam=expand("sorted_reads/{sample}.bam", sample=config["pe_samples"]),
        bai=expand("sorted_reads/{sample}.bam.bai", sample=config["pe_samples"])
    output:
        "../results/variant_calls/{sample}.vcf"
    run:
        #Check if pair-end or single-read
        if config["seq_type"] == "PE":
            shell(
                "bcftools mpileup -d 1000000 -f {input.ref} {input.bam} | "
                "bcftools call -mv > {output}")
        else:
            pass  # Skip this rule for single-read data
'''

rule gatk_mutect2_tumor_only:
    input:
        ref="../data/references/hg38.fa",
        bam="../results/sorted_reads/{sample}.bam",
        bai="../results/sorted_reads/{sample}.bam.bai",
        bed="../data/bed/OCAPlus.fixed.bed"
    output:
        "../results/variant_calls/{sample}.raw.vcf.gz"
    conda:
        "envs/gatk.yaml"
    shell:
        "gatk Mutect2 -R {input.ref} \
        -I {input.bam} \
        -O {output}"